runelite-client/src/main/java/net/runelite/client/ui/overlay/components/TooltipComponent.java

```diff
@@ -109,43 +140,77 @@ public Dimension render(Graphics2D graphics, Point parent)
 		{
 			lineX = textX;
 			final String line = lines[i];
-			final Matcher m = COLOR_SPLIT.matcher(line);
+			char[] chars = line.toCharArray();
 
 			int begin = 0;
-			while (m.find())
+			for (int j = 0; j < chars.length; j++)
 			{
-				// Draw text prior to color tag
-				String preText = line.substring(begin, m.start());
-				final TextComponent textComponent = new TextComponent();
-				textComponent.setText(preText);
-				textComponent.setPosition(new Point(lineX, textY + (i + 1) * textHeight - textDescent));
-				textComponent.setColor(nextColor);
-				textComponent.render(graphics, parent);
-
-				// Set color for next text part
-				if (m.group(1) == null)
+				if (chars[j] == '<')
 				{
-					// no color tag
-					nextColor = Color.WHITE;
+					TextComponent textComponent = new TextComponent();
+					textComponent.setColor(nextColor);
+					String text = line.substring(begin, j);
+					textComponent.setText(text);
+					textComponent.setPosition(new Point(lineX, textY + (i + 1) * textHeight - textDescent));
+					textComponent.render(graphics, parent);
+
+					lineX += metrics.stringWidth(text);
+
+					begin = j;
 				}
-				else
+				else if (chars[j] == '>')
 				{
-					// color tag
-					nextColor = Color.decode("#" + m.group(1));
+					String subLine = line.substring(begin + 1, j);
+
+					if (subLine.startsWith("col="))
+					{
+						String argument = subLine.substring(4);
+						nextColor = Color.decode("#" + argument);
+					}
+					else if (subLine.equals("/col"))
+					{
+						nextColor = Color.WHITE;
+					}
+					else if (subLine.startsWith("img="))
+					{
+						String argument = subLine.substring(4);
+						int iconId = Integer.parseInt(argument);
+						IndexedSprite modIcon = client.getModIcons()[iconId];
+						renderModIcon(graphics, lineX, textY + i * textHeight - textDescent, modIcon);
+						lineX += modIcon.getWidth();
+					}
+
+					begin = j + 1;
 				}
-				begin = m.end();
-				lineX += metrics.stringWidth(preText);
 			}
 
 			// Draw trailing text (after last tag)
-			final TextComponent textComponent = new TextComponent();
+			TextComponent textComponent = new TextComponent();
+			textComponent.setColor(nextColor);
 			textComponent.setText(line.substring(begin, line.length()));
 			textComponent.setPosition(new Point(lineX, textY + (i + 1) * textHeight - textDescent));
-			textComponent.setColor(nextColor);
 			textComponent.render(graphics, parent);
 		}
 
-
 		return new Dimension(tooltipWidth + OFFSET * 2, tooltipHeight + OFFSET * 2);
 	}
+
+	private void renderModIcon(Graphics2D graphics, int x, int y, IndexedSprite modIcon)
```

Comment:
Wouldn't converting the modIcon first to awt Image and then drawing that be faster than drawing it pixel-by-pixel here? Not sure how the Image/BufferedImage is implemented internally, but it's worth a shot i guess.

Reply:
I doubt it. Converting it to an Image takes computing power too.

Reply:
Yeah true, it would've to be cached somewhere to be, and doing that isn't worth it I guess.